Sources/SwiftParser/Parser.swift

```diff
@@ -629,3 +649,49 @@ extension Parser {
     )
   }
 }
+
+// MARK: Incremental Parsing
+extension Parser {
+  mutating func loadCurrentSyntaxNodeFromCache(for kind: SyntaxKind) -> Syntax? {
+    guard let parseTransition = self.parseTransition,
+      let parseNodeAffectRange = self.parseNodeAffectRange
+    else {
+      return nil
+    }
+    var lookUpHelper = IncrementalParseLookup(transition: parseTransition, nodeAffectRangeCollector: parseNodeAffectRange)
+    let currentOffset = self.lexemes.getOffsetToStart(self.currentToken)
+    if let node = lookUpHelper.lookUp(currentOffset, kind: kind) {
+      self.lexemes.advance(by: node.byteSize, currentToken: &self.currentToken)
+      return node
+    }
+
+    return nil
+  }
+
+  func registerNodeForIncrementalParse(node: RawSyntax, startToken: Lexer.Lexeme) {
+    if let parseNodeAffectRange {
+      parseNodeAffectRange.registerNodeForIncrementalParse(
+        node: node,
+        length: max(lookaheadFurthestOffset - self.lexemes.getOffsetToStart(startToken), node.byteLength + currentToken.byteLength)
+      )
```

Comment:
I use `node.byteLength + currentToken.byteLength` here since we may encounter some cases that we lookahead only when we are at some specific tokens. i.e. we lookahead to find valid trailing closure only when we are at `leftBrace`.